internal/pkg/agent/application/coordinator/state/state.go

```diff
@@ -246,79 +266,141 @@ func (cs *CoordinatorState) State() (s State) {
-	cs.sendState(cs.State())
+	// Try to send to stateChangedChan but don't block -- if its buffer is full
+	// then an update is already pending, and the changes we're reporting will
+	// be included.
+	select {
+	case cs.stateChangedChan <- struct{}{}:
+	default:
+	}
 }
 
-func (cs *CoordinatorState) sendState(state State) {
-	cs.subMx.RLock()
-	defer cs.subMx.RUnlock()
+func (cs *CoordinatorState) stateReporter() {
+	var subscribers []StateSubscription
+	// We support a variable number of subscribers and we don't want any of them
+	// to block each other or the CoordinatorState as a whole, so we need to
+	// listen with reflect.Select, which selects on an array of cases.
+	// Unfortunately this means we need to track the active select cases
+	// ourselves, including their position in the array.
+	//
+	// The ordering we use is:
+	// - first, the listener on subscribeChan
+	// - second, the listener on stateChangedChan
+	// - after that, two cases for each subscriber, in the same order as the
+	//   subscribers array: first its done channel, then its listener channel.
+	//
+	// All subscribers are included in the array of select cases even when some
+	// have already been updated, that way we don't need to worry about the
+	// order changing. Instead, subscribers that have already been updated have
+	// the listener channel of their select case set to nil.
+	selectCases := []reflect.SelectCase{
+		{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(cs.subscribeChan),
+		},
+		{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(cs.stateChangedChan),
+		},
+	}
+	const newSubscriberIndex = 0
+	const stateChangedIndex = 1
+	const firstSubscriberIndex = 2
+
+	currentState := cs.State()
+
+	// resetListeners is called when a state change notification arrives, to
+	// reactivate the listener channels of all subscribers and update their
+	// select case to the new state value.
+	resetListeners := func() {
+		currentState = cs.State()
+		for i, subscriber := range subscribers {
+			listenerIndex := firstSubscriberIndex + 2*i + 1
+			selectCases[listenerIndex].Chan = reflect.ValueOf(subscriber.ch)
+			selectCases[listenerIndex].Send = reflect.ValueOf(currentState)
+		}
+	}
+
+	// addSubscriber is a helper to add a new subscriber and its select
+	// cases to our lists.
+	addSubscriber := func(subscriber StateSubscription) {
+		subscribers = append(subscribers, subscriber)
+		selectCases = append(selectCases,
+			// Add a select case receiving from the subscriber's done channel
+			reflect.SelectCase{
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(subscriber.ctx.Done()),
+			},
+			// Add a select case sending to the subscriber's listener channel
+			reflect.SelectCase{
+				Dir:  reflect.SelectSend,
+				Chan: reflect.ValueOf(subscriber.ch),
+				Send: reflect.ValueOf(currentState),
+			})
+	}
 
-	send := func(sub *StateSubscription) {
-		t := time.NewTimer(time.Second)
-		defer t.Stop()
+	for {
```

Comment:
As far as I can tell, the lifetime of the coordinator is the same as the lifetime as the agent process so this never exiting is likely fine. Is that understanding correct?

Seems like the entrypoint that would create this follows from 
https://github.com/elastic/elastic-agent/blob/588e550e2d5f23942598ebacca4cfce4c721e6e6/internal/pkg/agent/application/application.go#L164

Reply:
Yes, that's my understanding as well. I figured it was ok to leave this dangling since the previous version did the same with its own helper goroutines (every listener previously spawned a separate goroutine that did nothing but wait for that listener's context to end, which is no longer needed in this version), and there's no obvious context bounding `CoordinatorState`'s lifecycle.